JIT: Add explicit successor for BBJ_COND false branch

amanasifkhalid opened the pull request:
Part of #93020. Currently, if a `BBJ_COND` block's `bbJumpDest` branch is not taken, control falls through into the next block. Because we plan to remove this fall-through requirement, `BBJ_COND` blocks need a pointer to their "not-taken" branch target, as the target may no longer be the next block. This new pointer, `bbNormalJumpDest`, may also be used in the future to represent the finally continuation of a `BBJ_CALLFINALLY/BBJ_ALWAYS` pair, once this pattern is consolidated into one block with no implicit fall-through via #95355.

For now, `bbNormalJumpDest` models current behavior by always pointing to the next block. Follow-up work will introduce behavioral changes by allowing `bbNormalJumpDest` to diverge from `bbNext`.

amanasifkhalid committed:
Introduce bbNormalJumpDest

amanasifkhalid committed:
Replace Next() with GetNormalJumpDest()

amanasifkhalid committed:
Replace NextIs() with HasNormalJumpTo()

amanasifkhalid committed:
Style

amanasifkhalid commented on src/coreclr/jit/block.h:
```diff
@@ -573,11 +576,18 @@ struct BasicBlock : private LIR::Range
 
     void SetNext(BasicBlock* next)
     {
+        // TODO: remove
```
I meant to delete this; will get rid of it in the next revision.

SingleAccretion commented on src/coreclr/jit/fgdiagnostic.cpp:
```diff
@@ -2957,6 +2957,13 @@ void Compiler::fgDebugCheckBBlist(bool checkBBNum /* = false */, bool checkBBRef
         maxBBNum = max(maxBBNum, block->bbNum);
 
+        // BBJ_COND's normal (false) jump target is expected to be the next block
+        // TODO: Allow bbNormalJumpDest to diverge from bbNext
```
A little tip for writing future `TODO`s: in ongoing work, it is handy to add a unique postfix to a `TODO`, e. g. `TODO-NoFallThrough`, s. t. they can be text-searched for easily afterwards.

amanasifkhalid replied:
Thank you for the tip; I realized this might be a better approach about halfway through... I'll update the TODOs I introduced to make them easier to Ctrl+F when I remove them in the next PR.

amanasifkhalid commented:
CC @dotnet/jit-contrib, @AndyAyersMS PTAL -- sorry for the wait on this. [No diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=492567&view=ms.vss-build-web.run-extensions-tab) except for TP impact.

amanasifkhalid committed:
Update TODO comments

amanasifkhalid committed:
Style

amanasifkhalid committed:
Add comment explaining assert

BruceForstall commented:
What is the terminology we want to use for the two BBJ_COND branches? With this PR we now have "bbNormalJumpDest" indicating the "false" path and "bbJumpDest" indicating the "true" path. Neither of these words seem satisfactory to me, especially "normal". It seems like we should have "bbFalseJumpDest" and "bbTrueJumpDest" for BBJ_COND. For cases with only a single destination (e.g., BBJ_ALWAYS), just "bbJumpDest" seems fine.

We over-emphasize the word "jump"; maybe we can remove it from all those places, instead having "bbFalseDest", "bbTrueDest", "bbDest", "bbSwtDest", "bbEhfDest", "bbKind", "BBKinds", etc.

And maybe "Dest" should instead be "Target": "bbFalseTarget", "bbTrueTarget", "bbTarget", "bbSwtTarget", "bbEhfTarget". I think this is my preferred option.

@dotnet/jit-contrib Comments?

SingleAccretion commented:
FWIW, here's the origin of the "normal" terminology: https://github.com/dotnet/runtime/pull/93772#discussion_r1367399266.

I agree it would be best to use "True/False", the problem was/is that we use `GetJumpDest` for the "true" path currently.

amanasifkhalid commented:
I agree this would be a good time to rename these members.

> It seems like we should have "bbFalseJumpDest" and "bbTrueJumpDest" for BBJ_COND.

For `bbTrueTarget`, are you suggesting we introduce a new `BasicBlock*` member to the union of jump targets that essentially functions as an alias for `bbJumpDest`, and then rename `bbNormalJumpDest` to `bbFalseTarget`? Or would you prefer a new jump kind altogether for `BBJ_COND`, like we do for `BBJ_SWITCH`? This new type would be a pointer to a struct of two `BasicBlock*` members: the true and the false targets. This would also come with new helper methods specific to `BBJ_COND` (like `SetJumpKindAndTarget`, etc.) for manipulating this new jump kind.

I concur with your other naming suggestions, @BruceForstall.

BruceForstall commented:
> For bbTrueTarget, are you suggesting we introduce a new BasicBlock* member to the union of jump targets that essentially functions as an alias for bbJumpDest, and then rename bbNormalJumpDest to bbFalseTarget? 

Yes, this is what I was thinking.

Yes, there would be new helpers that would assert on BBJ_COND and return/set the new union member.

This probably means various "switch" cases where BBJ_COND behavior re-uses the same "branch" behavior and `bbJumpDest` access as other kinds would have to be split out, to access the new member. Maybe that's ok anyway, and clearer. (And the native compilers hopefully will "re-merge" if possible in the generated code.)

amanasifkhalid committed:
Rename bbNormalJumpDest to bbFalseTarget

amanasifkhalid committed:
Rename bbJumpDest to bbTarget

AndyAyersMS commented:
I agree true/false makes more sense, as abstractly before block layout all control flow transfers are jumps.

Ultimately, I would like to see the blocks refer to `FlowEdge`s instead of directly to blocks, so for `BBJ_COND` we'd have say `TrueEdge` and `FalseEdge`, and we'd enhance `FlowEdge` to refer to the target block (currently implicit); but we could also have combo methods that then access the edge and then its target. But if/when we get to this point we can still have `BasicBlock*` valued `TrueTarget` and `FalseTarget` methods on a block, they'll just do a bit more work.

amanasifkhalid committed:
Rename SetJumpKindAndTarget variants

amanasifkhalid committed:
Rename bbJumpKind

BruceForstall commented:
> Ultimately, ...

I presume you're stating this as a long-term goal, and not something we should consider for this, or a short-term PR? Thus, we can evolve in that direction but still take name changes and block pointer implementation as proposed (if agreed to)?

amanasifkhalid committed:
Rename bbJumpOffs

amanasifkhalid committed:
Rename bbJumpSwt

amanasifkhalid committed:
Rename bbJumpEhf

amanasifkhalid committed:
Rename HasJumpTo

amanasifkhalid committed:
Rename method args, dspJumpKind

amanasifkhalid committed:
Introduce getters/setters/checkers for bbTrueTarget

AndyAyersMS commented:
> > Ultimately, ...
> 
> I presume you're stating this as a long-term goal, and not something we should consider for this, or a short-term PR? Thus, we can evolve in that direction but still take name changes and block pointer implementation as proposed (if agreed to)?

Yes. It's a subsequent step in the "plan" outlined in #93020.

amanasifkhalid committed:
Replace GetTarget with GetTrueTarget for BBJ_COND

amanasifkhalid committed:
Replace SetTarget with SetTrueTarget for BBJ_COND

amanasifkhalid committed:
Replace TargetIs with TrueTargetIs for BBJ_COND

amanasifkhalid committed:
Style

amanasifkhalid committed:
Fix comment

amanasifkhalid committed:
Fix bbNext/bbTarget access for BBJ_COND in BBSuccList

amanasifkhalid commented:
I've renamed the `BasicBlock` members @BruceForstall mentioned above, and added new helper methods for setting/accessing/checking `BBJ_COND`'s true and false successors. I didn't add a `BBJ_COND`-specific variant of `SetKindAndTarget` in this change, as that might be best reserved for follow-up work once the false successor can diverge from `bbNext`, and we need a way to explicitly set both the true and false targets -- would we like a separate definition of `SetKindAndTarget` for initializing `bbTrueTarget`?

I apologize for the size of this change; at the very least, it does not produce any asmdiffs locally.

jakobbotsch commented:
> would we like a separate definition of `SetKindAndTarget` for initializing `bbTrueTarget`?

I think splitting these overloads to separate methods would make sense -- e.g. `SetSwitch`, `SetCond` and so on.

amanasifkhalid committed:
Split out SetSwtKindAndTarget and SetEhfKindAndTarget

amanasifkhalid committed:
Add SetCondKindAndTarget

jakobbotsch commented on src/coreclr/jit/block.h:
```diff
@@ -606,98 +616,164 @@ struct BasicBlock : private LIR::Range
+        return (bbFalseTarget == target);
+    }
+
+    void SetCondKindAndTarget(BasicBlock* target)
```
This doesn't exactly roll off the tongue -- why not `SetCond`? Eventually we would hopefully end up with just `SetCond(BasicBlock* falseTarget, BasicBlock* trueTarget)`, `SetSwitch(BBswtDesc* swtTarget)` and so on.

amanasifkhalid replied:
I can shorten the names of all the `SetKind...` variants. Would you be ok with `SetKindAndTarget` being left as-is? I'm not sure how we can intuitively shorten that one while still conveying it's setting the jump kind and target.

jakobbotsch replied:
Yes, I think `SetKindAndTarget` makes sense -- it's sort of a low-level operation, while these other ones are more high-level.

amanasifkhalid committed:
Rename to SetCond

amanasifkhalid committed:
Rename to SetSwitch

amanasifkhalid committed:
Rename to SetEhf

amanasifkhalid committed:
Rename to GetSwitchTarget (for consistency)

AndyAyersMS reviewed (approved):
A few small suggestions. Feel free to defer those to a subsequent PR.

amanasifkhalid committed:
Rename GetSwitchTarget > GetSwitchTargets

amanasifkhalid committed:
Rename GetEhfTarget > GetEhfTargets

amanasifkhalid committed:
Rename bbSwtTarget > bbSwtTargets

amanasifkhalid committed:
Rename bbEhfTarget/SetEhfTarget

amanasifkhalid committed:
Add asserts

amanasifkhalid commented:
OSX arm64 NativeAOT failure looks unrelated.

jakobbotsch commented:
I hit asserts with `JitDump` enabled after this change in some simple code examples. For example
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static int abs(int num) => num < 0 ? -num : num;
```
hits
```scala
Assert failure(PID 21588 [0x00005454], Thread: 51020 [0xc74c]): Assertion failed 'KindIs(BBJ_COND)' in 'Program:abs(int):int' during 'If conversion' (IL size 9; hash 0x993542c0; FullOpts)

    File: C:\dev\dotnet\runtime4\src\coreclr\jit\block.h Line: 661
    Image: D:\dev\core_roots\6c7e6e2e50550737260e40f0b472158bc15a6f82\corerun.exe
```

when `DOTNET_JitDump=abs`. Can you take a look? (Also might be a good idea to run SPMI through with `JitDump=*`).

amanasifkhalid commented:
@jakobbotsch Thank you for letting me know, taking a look now...
